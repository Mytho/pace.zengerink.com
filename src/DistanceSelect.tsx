import React from 'react'
import NumberSelect from './NumberSelect'

export enum DistanceFragment {
  Kilometers = 'km',
  Meters = 'm',
}

interface DistanceSelectProps {
  value: number
  onChange: (value: number) => void
}

interface Distance {
  [key: string]: number
}

export default class DurationSelect extends React.Component<DistanceSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const distance = this.toDistance(this.props.value)
    distance[prop] = value
    this.props.onChange(this.toMeters(distance))
  }

  toMeters(distance: Distance): number {
    return distance[DistanceFragment.Kilometers] * 1000 + distance[DistanceFragment.Meters]
  }

  toDistance(meters: number): Distance {
    return {
      [DistanceFragment.Kilometers]: Math.floor(meters / 1000),
      [DistanceFragment.Meters]: meters % 1000,
    }
  }

  render() {
    const distance = this.toDistance(this.props.value)

    return (
      <div className="duration-select">
        <NumberSelect
          name={DistanceFragment.Kilometers}
          value={distance[DistanceFragment.Kilometers]}
          max="100"
          onChange={this.handleChange.bind(this, DistanceFragment.Kilometers)}
        />
        <span>.</span>
        <NumberSelect
          name={DistanceFragment.Meters}
          value={distance[DistanceFragment.Meters]}
          max="999"
          pad="3"
          onChange={this.handleChange.bind(this, DistanceFragment.Meters)}
        />
        <span>km</span>
      </div>
    )
  }
}
