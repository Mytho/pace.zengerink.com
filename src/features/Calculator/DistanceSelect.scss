@import './Calculator.scss';

.distance-select {
    @include calculator-row;
}