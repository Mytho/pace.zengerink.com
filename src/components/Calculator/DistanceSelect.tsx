import React from 'react'
import NumberSelect from './NumberSelect'
import './DistanceSelect.scss'

export enum DistanceFragment {
  Kilometers = 'km',
  Meters = 'm',
}

interface DistanceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface Distance {
  [key: string]: number
}

export default class DurationSelect extends React.Component<DistanceSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const distance = this.toDistance(this.props.value)
    distance[prop] = value
    this.props.onChange(this.toMeters(distance))
  }

  toMeters(distance: Distance): number {
    return distance[DistanceFragment.Kilometers] * 1000 + distance[DistanceFragment.Meters]
  }

  toDistance(meters: number): Distance {
    return {
      [DistanceFragment.Kilometers]: Math.floor(meters / 1000),
      [DistanceFragment.Meters]: meters % 1000,
    }
  }

  render() {
    const distance = this.toDistance(this.props.value)

    return (
      <div className="distance-select">
        <div className="distance-select__label">{this.props.label}</div>
        <div className="distance-select__input">
          <NumberSelect
            name={DistanceFragment.Kilometers}
            value={distance[DistanceFragment.Kilometers]}
            max="100"
            onChange={this.handleChange.bind(this, DistanceFragment.Kilometers)}
          />
          <span className="distance-select__separator">.</span>
          <NumberSelect
            name={DistanceFragment.Meters}
            value={distance[DistanceFragment.Meters]}
            max="999"
            pad="3"
            onChange={this.handleChange.bind(this, DistanceFragment.Meters)}
          />
          <span className="distance-select__value-label">km</span>
        </div>
      </div>
    )
  }
}
