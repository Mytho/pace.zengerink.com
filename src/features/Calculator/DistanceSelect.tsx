import React, { ReactElement } from 'react'
import NumberSelect from './NumberSelect'
import './DistanceSelect.scss'

type DistanceFragmentName = 'km' | 'm'

interface DistanceFragments {
  km: number
  m: number
}

interface DistanceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

const DistanceSelect = (props: DistanceSelectProps): ReactElement => {
  const toMeters = (fragments: DistanceFragments): number => fragments.km * 1000 + fragments.m

  const toDistanceFragments = (meters: number): DistanceFragments => ({
    km: Math.floor(meters / 1000),
    m: meters % 1000,
  })

  const handleChange = (fragmentName: DistanceFragmentName, value: number) => {
    const fragments = toDistanceFragments(props.value)
    fragments[fragmentName] = value
    props.onChange(toMeters(fragments))
  }

  const distanceFragments = toDistanceFragments(props.value)

  return (
    <div className="distance-select">
      <div className="distance-select__label">{props.label}</div>
      <div className="distance-select__input">
        <NumberSelect
          name="km"
          value={distanceFragments.km}
          max="100"
          onChange={(value) => handleChange('km', value)}
        />
        <span className="distance-select__separator">.</span>
        <NumberSelect
          name="m"
          value={distanceFragments.m}
          max="999"
          pad="3"
          onChange={(value) => handleChange('m', value)}
        />
        <span className="distance-select__value-label">m</span>
      </div>
    </div>
  )
}

export default DistanceSelect
