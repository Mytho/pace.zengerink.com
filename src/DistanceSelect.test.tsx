import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import DistanceSelect, { DistanceFragment } from './DistanceSelect'


describe('render', () => {
  it('renders correctly', () => {
    const onChange = jest.fn()
    const wrapper = shallow(<DistanceSelect value={3661} onChange={onChange} />)
    const json = shallowToJson(wrapper)
    expect(json).toMatchSnapshot()
  })
})

describe('onChange', () => {
  function testOnChange(name: string, newValue: number, expectedArgument: number) {
    const onChange = jest.fn()
    const wrapper = shallow(<DistanceSelect value={0} onChange={onChange} />)
    const select = wrapper.find(`[name="${name}"]`)
    expect(select.length).toBe(1)

    select.simulate('change', newValue)
    expect(onChange.mock.calls.length).toBe(1)
    expect(onChange.mock.calls[0][0]).toBe(expectedArgument)
  }

  it('is called when one of the selects changes', () => {
    testOnChange(DistanceFragment.Kilometers, 1, 1000)
    testOnChange(DistanceFragment.Meters, 2, 2)
  })
})