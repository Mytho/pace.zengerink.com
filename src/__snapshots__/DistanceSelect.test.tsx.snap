// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div
  className="duration-select"
>
  <NumberSelect
    max="100"
    name="km"
    onChange={[Function]}
    value={3}
  />
  <span>
    .
  </span>
  <NumberSelect
    max="999"
    name="m"
    onChange={[Function]}
    pad="3"
    value={661}
  />
  <span>
    km
  </span>
</div>
`;
